### PR TITLE
Kullback-Leibler statistic and framework for reliable testing of flaky tests

### PR DESCRIPTION
**Pull request by alxempirical**
Part I of a series...

- Compute KL divergence
- Test KL calculation
- Framework for measuring thresholds on flaky statistical tests

### CODE REVIEW

File: tests/kl.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    calculated. In general the mean and variance of log(P(x)) is not known to
+    be finite, but it will be for any distribution crosscat generates at the
+    moment, because they all have finite entropy. Hence the Central Limit
+    Theorem applies at some sample size, and this can in principle be used as a
+    rough guide to the precision of the estimate. In tests comparing the
+    univariate gaussians N(0,1) and N(0,2), it tended to have a visually
+    obvious bias for sample sizes below 100,000.
+
+    """
+
```
Comment (gregory-marton) [inline tests/kl.py]: Nit: please add explicit pass

**Comment by gregory-marton** [comment]
LGTM: merge at will.

**Comment by riastradh-probcomp** [comment]
Hmm...  I think we should have a little more clarity about what this means, concisely summarizable both for the author of a test and for the operator of jenkins, before proliferating more stochastic test frameworks.  (For the record, there is another outstanding pull request for a similar purpose, <https://github.com/probcomp/bayeslite/pull/393>, which also puts nondeterministic seed choice, and reporting of seeds for reproduction, in one place.)

**Comment by alxempirical** [comment]
I welcome feedback about anything in `threshold.py` which needs further clarification.

**Comment by riastradh-probcomp** [comment]
I haven't had time to look closely, but from a quick skim, I'm not clear on what the responsibilities of a test author are, nor the responsibilities of a text executor, nor the properties provided by the framework to the reader of test output.  Can you write a quick summary from those three perspectives?

I'm also not clear from the overall structure what any of the computation semantically provides -- there seems to be KL divergence involved (of what distributions? is that part of the test? is there some assumed beta-Bernoulli/beta-binomial distribution involved?) and Monte Carlo estimation (so the test framework itself is stochastic, even beyond running individual stochastic tests?).  It's split between four different classes and two different files, so a summary of how they fit together would be nice.

**Comment by alxempirical** [comment]
What these changes provide is coincidentally a simpler version of what @axch described in his [recent blog post](http://alexey.radul.name/ideas/2016/on-testing-probabilistic-programs/).  It's a framework for empirically determining the threshold on a statistic in order to place an upper bound on the probability of n independent draws of the statistic all being less than said threshold.  That way you can say guarantee that despite stochasticity the test suite is more reliable than, say, the system it's running on.

Test author makes a test statistic like [`kullback_leibler_check_statistic`](https://github.com/probcomp/bayeslite/pull/398/files#diff-00ac05b91a37fbb375b3b16d69fb40b2R26), and runs that through `threshold.compute_sufficiently_stringent_threshold` as shown in [`compute_kl_threshold`](https://github.com/probcomp/bayeslite/pull/398/files#diff-00ac05b91a37fbb375b3b16d69fb40b2R59), specifying the number of times the threshold needs to be violated for the test to have failed and the desired failure probability.  Test executor then [calls `threshold.check_generator`](https://github.com/probcomp/bayeslite/pull/398/files#diff-00ac05b91a37fbb375b3b16d69fb40b2R78) with a function which generates the statistic, the number of times it has to fail in a row, the threshold, and the probability bound.  First time the test exceeds the threshold, it's deemed to have passed.  If it's less than the threshold n times in a row, a [`MultipleTestStatisticFailures`](https://github.com/probcomp/bayeslite/pull/398/files#diff-2beaafb051d05225b07e979b48ee6829R206) exception is raised, [reporting](https://github.com/probcomp/bayeslite/pull/398/files#diff-2beaafb051d05225b07e979b48ee6829R177) the generator function, the threshold, the number of times the threshold has been violated, and the bound on the probability of this happening under normal behavior of the system under test.

**Comment by alxempirical** [comment]
BTW, if that clarifies anything the docstring to `threshold.py` left unclear, please let me know.  And the classes are three `namedtuple`s and an exception.  Very little complexity there.

**Comment by gregory-marton** [comment]
I plan to re-review and resolve the above shortly. For my curiosity, can you foreshadow what you plan for parts [2..n] of the series?

**Comment by alxempirical** [comment]
[Part II](https://github.com/alxempirical/bayeslite/commit/2e0a06c0e8656a224b2ab767a9c3087483e50c9c)  is under review by @madeleineth, and Part III is uncertain, but I intend it to work by generating data from the crosscat prior as in [this commit](https://github.com/alxempirical/bayeslite/blob/a22ea2728b8dc35137daf20b21d9c13aaf1b4de2/tests/crosscat_data.py), feeding it back into crosscat via a generalization to _n_ columns of the `BayesDBTestDistribution` from Part II, and comparing BQL `PROBABILITY`s to the pdf which results from constructing the crosscat data, using the KL diveregence function and the original sample data as the density approximation to integrate over.

**Comment by gregory-marton** [comment]
In keeping with preliminary thoughts around light review*, and because it looks well enough documented for now, LGTM again.

* https://docs.google.com/document/d/1IvLvAUzlViq9yQIYEMLl0KFaH2ZaTgeXQrbTgVSTW0M/edit#heading=h.t506ahffb9ko

**Comment by fsaad** [comment]
@gregory-marton the linked document is very helpful and useful, perhaps we should publicize it?

**Comment by gregory-marton** [comment]
Is a draft, comments are welcome, and we will agree on some version of it
and publicize it at eng meeting. I'm glad you find it useful!

Grem

Gregory Marton
617-858-0775
http://csail.mit.edu/~gremio/cal
On May 2, 2016 7:24 PM, "F Saad" <notifications@github.com> wrote:

@gregory-marton <https://github.com/gregory-marton> the linked document is
very helpful and useful, perhaps we should publicize it?

—
You are receiving this because you were mentioned.

Reply to this email directly or view it on GitHub
<https://github.com/probcomp/bayeslite/pull/398#issuecomment-216394090>

**Comment by alxempirical** [comment]
Can I have bits to read it, please?

**Pull request merged**